### PR TITLE
async-icq simapp & demo

### LINKED ISSUE #112: async-icq demo controller chain implementation should live in this repo
async-icq has the concept of a host chain (the chain where queries are received and processed) and a controller chain (the chain sending the query req and processing the ack). there is a demo that currently lives [here](https://github.com/strangelove-ventures/interchain-query-demo) but i think it would be more beneficial for app developers to have access to this demo in the ibc-apps repo. it would also make testing via an e2e test with ictest in this repo straightforward so that functionality for both the host and controller can be exercised in CI

### PR DESCRIPTION
**Pull request by Reecepbcups**
closes #112 

## summary
Adds host simapp & controller module example + CI test for async icq

PR is so large because of the x/ copy paste from the old repo

### CODE REVIEW

File: modules/async-icq/Makefile [hunk 1]
```diff
@@ -2,19 +2,17 @@
...
-ALL_GO_FILES := $(shell find . -regex ".*\.go$$" | grep -v pb.go)
 VERSION := $(shell echo $(shell git describe --always) | sed 's/^v//')
 COMMIT := $(shell git log -1 --format='%H')
+GOMOD := $(shell go list -m)
 LEDGER_ENABLED ?= true
 BINDIR ?= $(GOPATH)/bin
 BUILDDIR ?= $(CURDIR)/build
 SIMAPP = ./testing/simapp
-MOCKS_DIR = $(CURDIR)/tests/mocks
+MOCKS_DIR = $(CURDIR)/test/mock
```
Comment (jtieri) [inline modules/async-icq/Makefile:12]: do we still need this? i don't think we use mocks in async-icq anymore

File: modules/async-icq/testing/demo-simapp/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
+# interchainquerydemo
+**ICQ Demo** 
+
+## Disclaimer
+
+The following repository and [`x/interquery`](./x/interquery/) module serves as an example and is used to exercise the functionality of Interchain Accounts end-to-end for development purposes only.
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/README.md:6]: I know this was incorrect in the original repo, `Interchain Accounts` should be replaced with `async-icq`

File: modules/async-icq/testing/demo-simapp/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+## Overview 
+
+The following repository contains a basic example of an Interchain Queries module and serves as a developer guide for teams that wish to use interchain queries functionality.
+
+### Developer Documentation
+
+Interchain Queries developer docs can be found here
+
+https://github.com/strangelove-ventures/ibc-go/blob/feature/icq_implementation/modules/apps/icq/README.md
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/README.md:17]: we could reference the updated doc in this repo now, https://github.com/cosmos/ibc-apps/tree/main/modules/async-icq

File: modules/async-icq/testing/demo-simapp/app/app.go [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+	icqkeeper "github.com/cosmos/ibc-go/v3/modules/apps/icq/keeper"
+	icqtypes "github.com/cosmos/ibc-go/v3/modules/apps/icq/types"
+	"github.com/cosmos/ibc-go/v3/modules/apps/transfer"
+	ibctransferkeeper "github.com/cosmos/ibc-go/v3/modules/apps/transfer/keeper"
+	ibctransfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	ibc "github.com/cosmos/ibc-go/v3/modules/core"
+	ibcclient "github.com/cosmos/ibc-go/v3/modules/core/02-client"
+	ibcclientclient "github.com/cosmos/ibc-go/v3/modules/core/02-client/client"
+	ibcclienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	ibcporttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/app/app.go:83]: should all these imports be updated to use ibc-go v7?

File: modules/async-icq/testing/demo-simapp/app/app.go [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	"github.com/tendermint/tendermint/libs/log"
+	tmos "github.com/tendermint/tendermint/libs/os"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/tendermint/starport/starport/pkg/cosmoscmd"
+
+	interquerymodule "github.com/quasar-finance/interchain-query-demo/x/interquery"
+	interquerymodulekeeper "github.com/quasar-finance/interchain-query-demo/x/interquery/keeper"
+	interquerymoduletypes "github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/app/app.go:97]: these import paths should be updated to reference the `x/interquery` module from this same repo vs. the one once maintained in quasars repo

File: modules/async-icq/testing/demo-simapp/app/simulation_test.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	simulationtypes "github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+	"github.com/quasar-finance/interchain-query-demo/app"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/app/simulation_test.go:15]: should be updated to reference the `/app` package in this same repo vs the quasar repo

File: modules/async-icq/testing/demo-simapp/cmd/icqd-demo/main.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package main
+
+import (
+	"os"
+
+	svrcmd "github.com/cosmos/cosmos-sdk/server/cmd"
+	"github.com/quasar-finance/interchain-query-demo/app"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/cmd/icqd-demo/main.go:7]: should be updated to reference the `/app` package in this same repo

File: modules/async-icq/testing/demo-simapp/go.mod [hunk 8]
```diff
@@ -0,0 +1,30 @@
+module github.com/quasar-finance/interchain-query-demo
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/go.mod:1]: module name should be updated to reflect the current repo vs. quasars

File: modules/async-icq/testing/demo-simapp/go.mod [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	github.com/tendermint/tendermint v0.34.21
+	github.com/tendermint/tm-db v0.6.6
+	google.golang.org/genproto/googleapis/api v0.0.0-20230726155614-23370e0ffb3e
+	google.golang.org/grpc v1.57.0
+	gopkg.in/yaml.v2 v2.4.0
+)
+
+replace (
+	// used latest commit of branch v3.3.0-icq
+	github.com/cosmos/ibc-go/v3 => github.com/strangelove-ventures/ibc-go/v3 v3.0.0-20221014082552-99c8caa484af
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/go.mod:26]: i don't recall why we needed this replace statement, perhaps it is still necessary which at that point would negate my previous comment regarding bumping the ibc-go version

File: modules/async-icq/testing/demo-simapp/proto/buf.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+version: v1
+name: buf.build/quasar-finance/interchain-query-demo
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/proto/buf.yaml:2]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/proto/interquery/genesis.proto [hunk 11]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+package quasarfinance.interchainquerydemo.interquery;
+
+import "gogoproto/gogo.proto";
+import "interquery/params.proto";
+// this line is used by starport scaffolding # genesis/proto/import
+
+option go_package = "github.com/quasar-finance/interchain-query-demo/x/interquery/types";
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/proto/interquery/genesis.proto:8]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/proto/interquery/params.proto [hunk 12]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package quasarfinance.interchainquerydemo.interquery;
+
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/quasar-finance/interchain-query-demo/x/interquery/types";
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/proto/interquery/params.proto:6]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/proto/interquery/query.proto [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+package quasarfinance.interchainquerydemo.interquery;
+
+import "gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "cosmos/base/query/v1beta1/pagination.proto";
+import "interquery/params.proto";
+import "google/protobuf/any.proto";
+// this line is used by starport scaffolding # 1
+
+option go_package = "github.com/quasar-finance/interchain-query-demo/x/interquery/types";
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/proto/interquery/query.proto:11]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/proto/interquery/tx.proto [hunk 14]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+package quasarfinance.interchainquerydemo.interquery;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+// this line is used by starport scaffolding # proto/tx/import
+
+option go_package = "github.com/quasar-finance/interchain-query-demo/x/interquery/types";
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/proto/interquery/tx.proto:7]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/scripts/protocgen.sh [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+echo "Generating gogo proto code"
+cd proto
+
+# generate gogo proto code
+buf generate --template buf.gen.gogo.yaml 
+
+cd ..
+
+# move proto files to the right places
+cp -r github.com/quasar-finance/interchain-query-demo/x/interquery/types/* x/interquery/types/
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/scripts/protocgen.sh:14]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/testutil/keeper/interquery.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/store"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
+	typesparams "github.com/cosmos/cosmos-sdk/x/params/types"
+	ibckeeper "github.com/cosmos/ibc-go/v3/modules/core/keeper"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/keeper"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/testutil/keeper/interquery.go:15]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/testutil/network/network.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/cosmos/cosmos-sdk/simapp"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/cosmos/cosmos-sdk/testutil/network"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/tendermint/starport/starport/pkg/cosmoscmd"
+	tmrand "github.com/tendermint/tendermint/libs/rand"
+	tmdb "github.com/tendermint/tm-db"
+
+	"github.com/quasar-finance/interchain-query-demo/app"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/testutil/network/network.go:21]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+	"fmt"
+	// "strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	// "github.com/cosmos/cosmos-sdk/client/flags"
+	// sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query.go:13]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query.go [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"fmt"
+	// "strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	// "github.com/cosmos/cosmos-sdk/client/flags"
+	// sdk "github.com/cosmos/cosmos-sdk/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query.go:11]: could drop these commented out imports

File: modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query_params.go [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/client/cli/query_params.go:8]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/x/interquery/client/cli/tx.go [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	// "github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/client/cli/tx.go:11]: should be updated to reflect the current repo vs. quasars repo. could also remove the commented out import

File: modules/async-icq/testing/demo-simapp/x/interquery/genesis.go [hunk 22]
```diff
@@ -0,0 +1,36 @@
+package interquery
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/keeper"
+	"github.com/quasar-finance/interchain-query-demo/x/interquery/types"
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/genesis.go:6]: should be updated to reflect the current repo vs. quasars repo

File: modules/async-icq/testing/demo-simapp/x/interquery/types/types.go [hunk 23]
```diff
@@ -0,0 +1 @@
+package types
```
Comment (jtieri) [inline modules/async-icq/testing/demo-simapp/x/interquery/types/types.go:1]: i don't think this file is necessary anymore

**Review by jtieri — COMMENTED** [review]
I left a bunch of duplicate comments, i was going to point out everywhere that import paths could be updated but quickly realized that there are a ton of them so stopped. I think a find/replace all will turn up all of the locations they can be changed. Sorry about that 😛 

Other than the handful of comments i had regarding small things this looks good!

File: modules/async-icq/testing/demo-simapp/app/app.go (hunk 4 above)
Comment (Reecepbcups) [reply modules/async-icq/testing/demo-simapp/app/app.go:83]: This uses a fork so I kept it as is: https://github.com/strangelove-ventures/ibc-go/tree/v3.3.0-icq

for this PR I just copied it over directly. A future PR can update it, but not a requirement for this test

File: modules/async-icq/testing/demo-simapp/x/interquery/types/types.go (hunk 23 above)
Comment (Reecepbcups) [reply modules/async-icq/testing/demo-simapp/x/interquery/types/types.go:1]: resolved in 3babd4a4c754d3c1da0ae43806597e2cbf270a93

File: modules/async-icq/testing/demo-simapp/README.md (hunk 3 above)
Comment (Reecepbcups) [reply modules/async-icq/testing/demo-simapp/README.md]: 053a49f9ef48e70b3f49267db7cf95d8935bf620

File: modules/async-icq/testing/demo-simapp/README.md (hunk 2 above)
Comment (Reecepbcups) [reply modules/async-icq/testing/demo-simapp/README.md]: 053a49f9ef48e70b3f49267db7cf95d8935bf620

File: modules/async-icq/Makefile (hunk 1 above)
Comment (Reecepbcups) [reply modules/async-icq/Makefile]: b3ad311fbf30ee7473966074b0c2e30156eb4280

File: modules/async-icq/testing/demo-simapp/go.mod (hunk 9 above)
Comment (Reecepbcups) [reply modules/async-icq/testing/demo-simapp/go.mod:26]: per my above comment, still required for now. Can maybe change out in a future PR to IBC v7 if that adds teh required features

**Review by jtieri — APPROVED** [review]
great work here! 🚢

**Pull request merged**